pkg/capabilities/triggers/cron_trigger.go

```diff
@@ -0,0 +1,287 @@
+	capabilities.Validator[cronTriggerConfig, cronTriggerInput, capabilities.CapabilityResponse]
+	jobs      map[string]cronJob
+	lggr      logger.Logger
+	mu        sync.Mutex
+	scheduler gocron.Scheduler
+	stopCh    services.StopChan
+	wg        sync.WaitGroup
+}
+
+var _ capabilities.TriggerCapability = (*CronTriggerService)(nil)
+var _ services.Service = &CronTriggerService{}
+
+func NewCronTriggerService(lggr logger.Logger) *CronTriggerService {
+	s, err := gocron.NewScheduler()
+	if err != nil {
+		return nil
+	}
+	return &CronTriggerService{
+		CapabilityInfo: cronTriggerInfo,
+		Validator:      capabilities.NewValidator[cronTriggerConfig, cronTriggerInput, capabilities.CapabilityResponse](capabilities.ValidatorArgs{Info: cronTriggerInfo}),
+		jobs:           map[string]cronJob{},
+		lggr:           logger.Named(lggr, "CronTrigger"),
+		scheduler:      s,
+		stopCh:         make(services.StopChan),
+	}
+}
+
+func (cts *CronTriggerService) RegisterTrigger(ctx context.Context, req capabilities.CapabilityRequest) (<-chan capabilities.CapabilityResponse, error) {
+	cts.mu.Lock()
+	defer cts.mu.Unlock()
+
+	config, err := cts.ValidateConfig(req.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	triggerID := getTriggerID(config.TriggerID, req.Metadata.WorkflowID)
+	if _, ok := cts.jobs[triggerID]; ok {
+		return nil, fmt.Errorf("triggerId %s already registered", triggerID)
+	}
+
+	job := gocron.CronJob(config.Schedule, true)
+
+	cronParser := cron.NewParser(
+		cron.SecondOptional | cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow,
+	)
+	schedule, err := cronParser.Parse(config.Schedule)
+	if err != nil {
+		return nil, err
+	}
+
+	t := gocron.NewTask(
+		cts.makeProcess(triggerID, schedule),
+	)
+
+	j, err := cts.scheduler.NewJob(
+		job,
+		t,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ch := make(chan capabilities.CapabilityResponse, defaultSendChannelBufferSize)
+	cts.jobs[triggerID] = cronJob{
+		ch:  ch,
+		job: j,
+	}
+	cts.lggr.Debugw("RegisterTrigger", "triggerId", triggerID, "jobId", j.ID())
+	return ch, nil
+}
+
+func (cts *CronTriggerService) makeProcess(triggerId string, schedule cron.Schedule) func() {
+	return func() { cts.process(triggerId, schedule) }
+}
+
+func (cts *CronTriggerService) process(triggerId string, schedule cron.Schedule) {
+	cts.mu.Lock()
+	defer cts.mu.Unlock()
+
+	now := time.Now()
+	scheduledExecutionTime := schedule.Next(now)
+	// if scheduledExecutionTime is before now, get the next one
+	// this can be due to time sync / sleep
+	for {
+		if scheduledExecutionTime.Before(now) {
+			scheduledExecutionTime = schedule.Next(now.Add(time.Second))
+		} else {
+			break
+		}
+	}
+
+	// timestamp in ns precision, to show the discrepancy between the scheduled time and the actual time
+	// we need scheduled time to generate a deterministic triggerEventId
+	timestampNs := time.Now().UTC().UnixNano()
+	// format to ISO 8601
+	timestampMsIso8601 := formatUnixNanoToISO8601(timestampNs)
```

Comment:
This feels overly cumbersome: can we just do `time.Now().Format(time.RFC3339Nano)`?

Reply:
Gets simplified with the Trigger interface using a direct timestamp